batchflow/models/tf/layers/core.py

```diff
@@ -50,25 +50,71 @@ def __call__(self, inputs):
 
 
 
-class Dropout(Layer):
-    """ Wrapper for dropout layer. """
-    def __init__(self, dropout_rate, **kwargs):
+class BaseDropout(Layer):
+    """ Base class for dropout layers.
+
+    Parameters
+    ----------
+    dropout_prob : float, tf.Tensor, callable
+        If float or Tensor, then fraction of the input units to drop.
+        If callable, then function to be called on `global_step`. Must return tensor of size 1.
+    multisample: bool, number, sequence
+        If evaluates to True, then batch is split into multiple parts,
+        dropout applied to each of them separately and then parts are concatenated back.
+
+        If True, then batch is split evenly into two parts.
+        If integer, then batch is split evenly into that number of parts; must be a divisor of batch size.
+        If float, then batch is split into parts of `multisample` and `1 - multisample` sizes.
+        If sequence of ints, then batch is split into parts of given sizes. Must sum up to the batch size.
+        If sequence of floats, then each float means proportion of sizes in batch and must sum up to 1.
+    """
+    def __init__(self, dropout_rate, multisample=False, global_step=None, **kwargs):
         self.dropout_rate = dropout_rate
+        self.global_step = global_step
+        self.multisample = multisample
         self.kwargs = kwargs
 
     def __call__(self, inputs, training):
-        return K.Dropout(rate=self.dropout_rate, **self.kwargs)(inputs, training)
+        #pylint: disable=singleton-comparison
+        if callable(self.dropout_rate):
+            step = tf.cast(self.global_step, dtype=tf.float32)
+            self.dropout_rate = self.dropout_rate(step)
+        d_layer = self.LAYER(rate=self.dropout_rate, **self.kwargs)
+
+        if self.multisample is not False:
+            if self.multisample is True:
+                self.multisample = 2
+            elif isinstance(self.multisample, float):
+                self.multisample = [self.multisample, 1 - self.multisample]
+
+            if isinstance(self.multisample, int):
+                sizes = self.multisample
+            elif isinstance(self.multisample, (tuple, list)):
+                if all([isinstance(item, int) for item in self.multisample]):
+                    sizes = self.multisample
+                elif all([isinstance(item, float) for item in self.multisample]):
+                    batch_size = tf.cast(tf.shape(inputs)[0], dtype=tf.float32)
+                    sizes = tf.convert_to_tensor([batch_size*item for item in self.multisample])
+                    sizes = tf.cast(tf.math.round(sizes), dtype=tf.int32)
+            elif isinstance(self.multisample, tf.Tensor):
+                sizes = self.multisample
+
+            splitted = tf.split(inputs, sizes, axis=0, name='mdropout_split')
```

Comment:
The capability to route different batch items to different dropouts is great.
But I reckon we also need the whole batch to go through each dropout, and after that combine is needed.